src/components/tree/tree.tsx

```diff
@@ -22,12 +26,71 @@ export const spaceBetweenTrees = 100;
 export const lhsTreeId = "LEFT";
 const rhsTreeId = "RIGHT";
 
-class Tree extends React.Component {
-  constructor(props) {
+export interface TreeComponentOwnProps {
+  dispatch: AppDispatch
+  height: number
+  width: number
+}
+
+export interface TreeComponentProps extends WithTranslation, TreeComponentStateProps, TreeComponentOwnProps {}
+
+// This is duplicated from RootState, but good to be explicit about what's
+// expected here.
+export interface TreeComponentStateProps {
+  animationPlayPauseButton: "Play" | "Pause"
+  canRenderBranchLabels: boolean
+  colorBy: string
+  colorByConfidence: boolean
+  colorings: any
+  colorScale: ColorScale
+  dateMaxNumeric: number
+  dateMinNumeric: number
+  distanceMeasure: any
+  explodeAttr: string
+  filters: Record<string, Array<{ value: string, active: boolean }>>
+  focus: boolean
+  genomeMap: any
+  layout: Layout
+  narrativeMode: boolean
+  panelsToDisplay: string[]
+  quickdraw: boolean
+  scatterVariables: ScatterVariables
+  selectedBranchLabel: string
+  selectedNode: SelectedNode | null
+  showAllBranchLabels: boolean
+  showOnlyPanels: boolean
+  showTangle: boolean
+  showTreeToo: boolean
+  temporalConfidence: TemporalConfidence
+  tipLabelKey: string | symbol
+  tree: TreeState
+  treeToo: TreeTooState
+}
+
+export interface TreeComponentState {
+  hoveredNode?: {
+    node: PhyloNode
+    isBranch: boolean
+  } | null
+  tree?: PhyloTreeType | null
+  treeToo?: PhyloTreeType | null
```

Comment:
I've replaced e20a865b18d78461b20542d41e07df2cff6cd7bf with 7d6613315cbc0cca4610d3e0c83f818b9dbffa41. However, I'm wondering if something like bdc91da7ac2e4816b27738b1e86fa52495fa8063 is a better solution. I can't tell any difference from some basic testing.

Reply:
https://github.com/nextstrain/auspice/commit/7d6613315cbc0cca4610d3e0c83f818b9dbffa41 is identical to https://github.com/nextstrain/auspice/commit/bdc91da7ac2e4816b27738b1e86fa52495fa8063, no? What's the advantage of this approach, React's `setState` will [shallow merge the provided object](https://react.dev/reference/react/Component#setstate) anyway. 

As for https://github.com/nextstrain/auspice/commit/e20a865b18d78461b20542d41e07df2cff6cd7bf, using optional properties like `{tree?: PhyloTree, treeToo?: PhyloTree}` isn't ideal because if hovered node is set then it will always have `tree` on it (it's not optional). As for `hoveredNode?: ...` vs `hoveredNode: type | null` they seem pretty similar for me.

Reply:
Sorry, I copied the wrong SHA. 9549dfbb63295ebd7b01efc6cd5c9a118d9f84b8 vs bdc91da7ac2e4816b27738b1e86fa52495fa8063

Reply:
Oh, I didn't realize React does shallow merge. I think there's an issue with the type definition of `setState` then... there should be no need for either of the two changes.

Reply:
I'm still not sure I understand

> https://github.com/nextstrain/auspice/commit/bdc91da7ac2e4816b27738b1e86fa52495fa8063 update state instead of replacing with a partial

`setState` will never replace the state with the provided state it'll shallow merge it in, so we don't have to do that merge. I.e. `this.setState({...this.state, ...newState})` is equivalent to `this.setState(newState)` (but slower and less clear).

Reply:
There is a StackOverflow thread on this: [allow typescript compiler to call setState on only one react state property](https://stackoverflow.com/q/37300933)

The answer says:

> If you still have this error then you might want to consider updating your react definitions.

I just checked and we are using `@types/react` version 17.0.39 ([published in 2022](https://www.npmjs.com/package/@types/react/v/17.0.39)):

https://github.com/nextstrain/auspice/blob/9549dfbb63295ebd7b01efc6cd5c9a118d9f84b8/package-lock.json#L17434-L17435

Looks like the type has changed since that answer. It is now:

```ts
// We MUST keep setState() as a unified signature because it allows proper checking of the method return type.
// See: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/18365#issuecomment-351013257
// Also, the ` | S` allows intellisense to not be dumbisense
setState<K extends keyof S>(
    state: ((prevState: Readonly<S>, props: Readonly<P>) => (Pick<S, K> | S | null)) | (Pick<S, K> | S | null),
    callback?: () => void
): void;
```

[This answer](https://stackoverflow.com/a/55824499) (less popular but more recent) recommends `setState<never>()`. It seems the most appropriate for our use case.

Reply:
Using `setState<never>()` in 7cefd6c545a2dff13e9c17536db11f9b37efb7d1

Reply:
> Edit: this message was written prior to reading the above response

This happened to me too. I think GitHub has real-time updates on issues but not PR comment threads. Unfortunate.